utils.go

```diff
@@ -82,17 +83,39 @@ func GetRootlessRuntimeDir(rootlessUID int) (string, error) {
 	return path, nil
 }
 
+func getInitCommand() (string, error) {
+	file, err := os.Open("/proc/1/comm")
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+	buf, err := ioutil.ReadAll(file)
+	if err != nil {
+		return "", err
+	}
+	return string(buf), nil
+}
+
 func getRootlessRuntimeDir(rootlessUID int) (string, error) {
 	runtimeDir, err := homedir.GetRuntimeDir()
 	if err == nil {
 		return runtimeDir, nil
 	}
-	tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
+	initCommand, err := getInitCommand()
+	var tmpDir string
+	if err == nil && initCommand == "systemd" {
+		// Systemd maintains lingering mode for /run/user/X.
+		// It is up to user whether this directory will survive logout.
+		tmpDir = fmt.Sprintf("/run/user/%d", rootlessUID)
+	} else {
+		// Other systems follows standard https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html.
+		// /run/user/X can't survive logout, so it is possible to use temp directory only.
+		tmpDir = fmt.Sprintf("%s/storage-user-%d", os.TempDir(), rootlessUID)
+	}
 	st, err := system.Stat(tmpDir)
 	if err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000 {
 		return tmpDir, nil
 	}
-	tmpDir = fmt.Sprintf("%s/%d", os.TempDir(), rootlessUID)
 	if err := os.MkdirAll(tmpDir, 0700); err != nil {
```

Comment:
This is insecure: an attacker can create a directory with this predictable name, `MkdirAll` will succeed, and the attacker can then attack the use of the runtime directory (e.g. by placing symlinks inside it, or replacing the directory itself.

(Using `os.Mkdir` _should_ be safe even in presence of other files or symlinks, as long as errors are not ignored. But please have a test for that. (Really, this function would benefit a 100% code coverage testing against various attacks — or even instrumentation to explicitly pause it in various points, like above this line, to let the “attacking” test always win a race.)

(If it were up to me, I would delete most of the fallback code in here and just expect `/run/user` and/or $HOME to work. OTOH I haven’t looked into what functionality that would break, so I’m clearly not qualified to make such a recommendation.)